src/utils/car.js

```diff
@@ -33,20 +37,49 @@ const hashes = {
 export async function validateBody (body) {
   const carBlockIterator = await CarBlockIterator.fromIterable(body)
   for await (const { cid, bytes } of carBlockIterator) {
-    if (!codecs[cid.code]) {
-      throw new Error(`Unexpected codec: 0x${cid.code.toString(16)}`)
-    }
-    if (!hashes[cid.multihash.code]) {
-      throw new Error(`Unexpected multihash code: 0x${cid.multihash.code.toString(16)}`)
-    }
-
-    // Verify step 2: if we hash the bytes, do we get the same digest as reported by the CID?
-    // Note that this step is sufficient if you just want to safely verify the CAR's reported CIDs
-    const hash = await hashes[cid.multihash.code].digest(bytes)
-    if (toHex(hash.digest) !== toHex(cid.multihash.digest)) {
-      throw new Error('Hash mismatch')
-    }
+    await verifyBlock(cid, bytes)
   }
 
   return true
 }
+
+/**
+ * Verifies a block
+ *
+ * @param {CID} cid
+ * @param {Uint8Array} bytes
+ */
+export async function verifyBlock (cid, bytes) {
```

Comment:
@rvagg do we have something like this in the JS ecosystem yet?https://github.com/ipfs/boxo/blob/303595bcdba7c4ff1e4186a8089a550df6da9d28/verifcid/cid.go#L22

Seems like it could be helpful as it's needed to prevent certain classes of attacks

Reply:
much tricker in JS because we don't just bundle a bunch of "standard" hashers, instead leaving it up to the user to maintain the list they care about

this continues a conversation here https://github.com/ipld/js-car/issues/123, and here https://github.com/multiformats/js-multiformats/issues/253

basically we're in need of an abstraction where the user plugs in the hashers they would like to verify, but for now we don't have a great answer for it unfortunately. I was hoping @gozala would find some time to work on this, maybe it can happen along with some of the ongoing work around hasher byte array reuse.

Reply:
verifcid is a much cheaper kind of wrapper. There aren't piles of code and dependencies needed here, it's just:
- CID parsing (you already have this from js-multiformats)
- An allowlist wrapper
- If you want to bundle sane defaults then it's manually defining a handful of constants

The point of it is to make sure:
1. You don't accidentally use hash functions that are problematic (like MD5) which is less of a problem in JS due to the need to manually include hashers everywhere
2. You don't handle multihash lengths that when combined with the given hash function make it insecure (e.g. a SHA2-256 hash truncated to 1 byte is not a secure hash function)

It's easy enough to add the basics here by checking if for the two supported hash functions (SHA2 and blake2b256) that the multihashes have digests with length >= 20 bytes, but it'd probably be nice to not have users repeat this logic in multiple places.

Reply:
added checks for minimum and maximum hash length